File: circuits/dleq_different_curves.py
import argparse
import hashlib
import os
from fastecdsa import curve, point
import pysodium
from binascii import hexlify, unhexlify

# secp256k1 curve parameters
SECP256K1 = curve.secp256k1
N1 = SECP256K1.q  # Order: 115792089237316195423570985008687907852837564279074904382605163141518161494337

# Ed25519 curve parameters
N2 = 7237005577332262213973186563042994240857116359379907606001950938285454250989  # Order

def hash_points(*points):
    """Compute SHA-256 hash of concatenated point encodings."""
    concatenated = b''.join(point if isinstance(point, bytes) else point.to_bytes(33, 'big') for point in points)
    return int.from_bytes(hashlib.sha256(concatenated).digest(), 'big') % N2

def generate_dleq_proof(a, B, G1):
    """Generate a DLEQ proof for A = a*G1 (secp256k1) and C = a*B (Ed25519)."""
    # Validate inputs
    if a <= 0 or a >= N1 or a >= N2:
        raise ValueError("Scalar a must be in [1, min(n1, n2))")
    if len(B) != 32:
        raise ValueError("Point B must be a 32-byte Ed25519 point")
    # if not pysodium.crypto_core_ed25519_is_valid_point(B):
    #     raise ValueError("Point B must be a valid Ed25519 point")

    # Compute A = a * G1 (secp256k1)
    A = a * G1
    A_bytes = A.to_bytes(33, 'big')  # Compressed point

    # Compute C = a * B (Ed25519)
    # pysodium expects little-endian scalar
    a_bytes = a.to_bytes(32, 'little')
    C = pysodium.crypto_scalarmult(a_bytes, B)

    # Generate proof
    k = int.from_bytes(os.urandom(32), 'big') % N2  # Random scalar
    kG = (k * G1).to_bytes(33, 'big')  # k * G1 (secp256k1)
    k_bytes = k.to_bytes(32, 'little')
    kB = pysodium.crypto_scalarmult(k_bytes, B)  # k * B (Ed25519)

    # Compute challenge c = H(kG || kB || A || B || C)
    c = hash_points(kG, kB, A_bytes, B, C)

    # Compute response r = k - c * a mod n2
    r = (k - c * a) % N2

    return {"c": c, "r": r}, A_bytes, C

def verify_dleq_proof(A, B, C, proof, G1):
    """Verify a DLEQ proof."""
    c, r = proof["c"], proof["r"]

    # Validate inputs
    if r >= N2 or c >= N2:
        raise ValueError("Proof scalars out of range")
    try:
        A_point = point.Point.from_bytes(A, SECP256K1)
    except ValueError:
        raise ValueError("Point A must be valid on secp256k1")
    if len(B) != 32 or len(C) != 32:
        raise ValueError("Points B and C must be 32-byte Ed25519 points")
    # if not pysodium.crypto_core_ed25519_is_valid_point(B) or not pysodium.crypto_core_ed25519_is_valid_point(C):
    #     raise ValueError("Points B and C must be valid Ed25519 points")

    # Compute R1 = r * G1 + c * A (secp256k1)
    A_point = point.Point.from_bytes(A, SECP256K1)
    R1 = r * G1 + c * A_point
    R1_bytes = R1.to_bytes(33, 'big')

    # Compute R2 = r * B + c * C (Ed25519)
    r_bytes = r.to_bytes(32, 'little')
    c_bytes = c.to_bytes(32, 'little')
    rB = pysodium.crypto_scalarmult(r_bytes, B)
    cC = pysodium.crypto_scalarmult(c_bytes, C)
    # Ed25519 point addition
    R2 = pysodium.crypto_core_ed25519_add(rB, cC)

    # Compute challenge c' = H(R1 || R2 || A || B || C)
    c_prime = hash_points(R1_bytes, R2, A, B, C)

    return c == c_prime

def main():
    parser = argparse.ArgumentParser(description="Run DLEQ protocol across secp256k1 and Ed25519")
    parser.add_argument("-a", "--scalar", type=int, required=True, help="Secret scalar a")
    parser.add_argument("-b", "--point-b", type=str, required=True, help="Point B on Ed25519 (hex)")
    args = parser.parse_args()

    try:
        a = args.scalar
        B = unhexlify(args.point_b)
        G1 = SECP256K1.G  # secp256k1 base point

        # Generate proof
        proof, A, C = generate_dleq_proof(a, B, G1)
        print("Proof:", {"c": str(proof["c"]), "r": str(proof["r"])})
        print("A (secp256k1, hex):", hexlify(A).decode())
        print("C (Ed25519, hex):", hexlify(C).decode())

        # Verify proof
        is_valid = verify_dleq_proof(A, B, C, proof, G1)
        print("Proof valid:", is_valid)
    except Exception as e:
        print("Error:", str(e))
        exit(1)

if __name__ == "__main__":
    main()

File: circuits/scripts/dleq_different_curves.js
// const { secp256k1, ed25519 } = require('@noble/curves');
const { secp256k1 } = require('@noble/curves/secp256k1');
const { ed25519 } = require('@noble/curves/ed25519');
const { sha256 } = require('@noble/hashes/sha256');
const { program } = require('commander');

// Function to generate a DLEQ proof across two curves
function generateDLEQProof(a, G2, G1) {
    // Validate inputs
    const n1 = BigInt(secp256k1.CURVE.n.toString()); // Order of secp256k1
    const n2 = BigInt(ed25519.CURVE.n.toString()); // Order of Curve25519
    if (n1 < n2) {
        throw new Error('n2 != min(n1, n2))');
    }
    if (a <= 0n || a >= n1 || a >= n2) {
        throw new Error('Scalar a must be in [1, min(n1, n2))');
    }

    // Compute A = a * G1 (secp256k1) and C = a * G2 (Curve25519)
    const A = G1.multiply(a);
    const C = G2.multiply(a);

    // Generate proof
    const k = BigInt('1');//BigInt('0x' + Buffer.from(sha256(Math.random().toString())).toString('hex')) % n2; // Random scalar
    const kG = G1.multiply(k); // k * G1
    const kB = G2.multiply(k); // k * B

    // Compute challenge c = H(kG || kB || A || B || C)
    const cHash = sha256(Buffer.concat([
        Buffer.from(kG.toHex(), 'hex'),
        Buffer.from(kB.toHex(), 'hex'),
        Buffer.from(A.toHex(), 'hex'),
        Buffer.from(G2.toHex(), 'hex'),
        Buffer.from(C.toHex(), 'hex')
    ]));
    const c = BigInt('0x' + Buffer.from(cHash).toString('hex')) % n2;

    // Compute response r = k - c * a mod n2
    var r = (k - c * a) % n2;
    if (r < 0n) r += n2;
    console.log('r:', r);
    console.log('n2:', n2);

    return { proof: { c, r }, A, C };
}

// Function to verify a DLEQ proof
function verifyDLEQProof(A, G2, C, proof, G1) {
    const n1 = BigInt(secp256k1.CURVE.n.toString());
    const n2 = BigInt(ed25519.CURVE.n.toString());
    const { c, r } = proof;

    // Validate inputs
    if (r >= n2 || c >= n2) {
        throw new Error('Proof scalars out of range');
    }

    // Compute R1 = r * G1 + c * A (secp256k1)
    const rG = G1.multiply(r);
    const cA = A.multiply(c);
    const R1 = rG.add(cA);

    // Compute R2 = r * G2 + c * C (Curve25519)
    const rB = G2.multiply(r);
    const cC = C.multiply(c);
    const R2 = rB.add(cC);

    // Compute challenge c' = H(R1 || R2 || A || B || C)
    const cPrimeHash = sha256(Buffer.concat([
        Buffer.from(R1.toHex(), 'hex'),
        Buffer.from(R2.toHex(), 'hex'),
        Buffer.from(A.toHex(), 'hex'),
        Buffer.from(G2.toHex(), 'hex'),
        Buffer.from(C.toHex(), 'hex')
    ]));
    const cPrime = BigInt('0x' + Buffer.from(cPrimeHash).toString('hex')) % n2;
    console.log('c:', c);
    console.log('cPrime:', cPrime);

    // Verify c == c'
    return c === cPrime;
}

// // Command line interface
// program
//     .version('1.0.0')
//     .description('Run DLEQ protocol across secp256k1 and Curve25519')
//     .requiredOption('-a, --scalar <decimal>', 'Secret scalar a')
//     .requiredOption('-b, --point-b <hex>', 'Point B on Curve25519 (hex)')
//     .action((options) => {
        // try {
            const a = BigInt("12345678901234567890");
            const G2 = ed25519.ExtendedPoint.fromHex("5866666666666666666666666666666666666666666666666666666666666666");
            console.log('B (Curve25519):', G2);
            const G1 = secp256k1.ProjectivePoint.BASE;

            // Generate proof
            const { proof, A, C } = generateDLEQProof(a, G2, G1);
            console.log('Proof:', { c: proof.c.toString(), r: proof.r.toString() });
            console.log('A (secp256k1):', A);
            console.log('C (Curve25519):', C);

            // Verify proof
            const isValid = verifyDLEQProof(
                A,
                G2,
                C,
                proof,
                G1
            );
            console.log('Proof valid:', isValid);
        // } catch (error) {
        //     console.error('Error:', error.message);
        //     process.exit(1);
        // }
//     });

// program.parse(process.argv);